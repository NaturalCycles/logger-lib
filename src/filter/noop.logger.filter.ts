import { LoggerFilter, LogObject } from '../logger.model'

export class NoopLoggerFilter implements LoggerFilter {
  filter (logObject: LogObject): LogObject {
    return logObject
  }
}
