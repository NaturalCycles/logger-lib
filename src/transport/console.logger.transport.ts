import { LOG_LEVEL, LoggerTransport, LogObject } from '../logger.model'

const CONSOLE_METHOD_MAP = {
  [LOG_LEVEL.DEBUG]: 'debug',
  [LOG_LEVEL.INFO]: 'log',
  [LOG_LEVEL.WARN]: 'warn',
  [LOG_LEVEL.ERROR]: 'error',
}

const DEF_METHOD = 'log'

export class ConsoleLoggerTransport implements LoggerTransport {
  send (o: LogObject): void {
    console[CONSOLE_METHOD_MAP[o.level] || DEF_METHOD](...o.args)
  }
}
