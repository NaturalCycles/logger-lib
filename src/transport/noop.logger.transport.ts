import { LoggerTransport, LogObject } from '../logger.model'

export class NoopLoggerTransport implements LoggerTransport {
  send (logObject: LogObject): void {
    // no-op, as advertised:)
  }
}
