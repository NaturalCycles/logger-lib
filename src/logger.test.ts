import { NoopLoggerFilter } from './filter/noop.logger.filter'
import { LOG_LEVEL, Logger, TAG_DEFAULT } from './logger.model'
import { LoggerService } from './logger.service'
import { ConsoleLoggerTransport } from './transport/console.logger.transport'
import { NoopLoggerTransport } from './transport/noop.logger.transport'

const noopTransport = new NoopLoggerTransport()
const transportMock = jest.spyOn(noopTransport, 'send')

const consoleTransport = new ConsoleLoggerTransport()
const consoleTransportMock = jest.spyOn(consoleTransport, 'send')

const noopFilter = new NoopLoggerFilter()
const noopFilterMock = jest.spyOn(noopFilter, 'filter')

beforeEach(() => {
  jest.resetAllMocks()
})

describe('minimal logger with debug level', () => {
  const logService = new LoggerService({
    transports: [noopTransport],
    tagLevels: {
      [TAG_DEFAULT]: LOG_LEVEL.DEBUG,
    },
  })

  test('basic api', () => {
    const log = logService.getLogger()
    callBasicAPI(log)
    expect(transportMock).toMatchSnapshot()
  })

  test('basic api with tag1', () => {
    const log = logService.getLogger().tag('tag1')
    callBasicAPI(log)
    expect(transportMock).toMatchSnapshot()
  })

  test('deduplicate tags', () => {
    const log = logService.getLogger().tag('tag1')
    log.tag('tag1').info('hello')
    expect(transportMock).toMatchSnapshot()
  })
})

describe('minimal logger with default levels', () => {
  const logService = new LoggerService({
    transports: [noopTransport],
  })

  test('basic api', () => {
    const log = logService.getLogger()
    callBasicAPI(log)
    // should ignore debug level
    expect(transportMock).toMatchSnapshot()
  })

  test('basic meta', () => {
    const log = logService.getLogger()
    log.meta({ a: 'a', n: 1 }).info('hello')
    expect(transportMock).toMatchSnapshot()
  })
})

describe('tag1:WARN', () => {
  const logService = new LoggerService({
    transports: [noopTransport],
    tagLevels: {
      tag1: LOG_LEVEL.WARN,
    },
  })

  test('tag1 to be filtered out', () => {
    const log = logService.getLogger()
    log.info('hello_notag') // pass
    log.tag('tag1').debug('hello_tag1')
    log.tag('tag1').info('hello_tag1')
    log.tag('tag1', 'tag2').info('hello_tag1')
    expect(transportMock).toHaveBeenCalledTimes(1)
  })

  test('tag1 to be filtered out with default tag1', () => {
    const log = logService.getLogger().tag('tag1')
    log.info('hello_notag')
    log.tag('tag1').info('hello_tag1') // will be deduplicated
    log.tag('tag2').info('hello_tag1') // still filtered out because of tag1
    log.warn('hello_warn') // will pass
    expect(transportMock).toHaveBeenCalledTimes(1)
  })
})

describe('TAG_DEFAULT:WARN;tag1:INFO', () => {
  const logService = new LoggerService({
    transports: [noopTransport],
    tagLevels: {
      [TAG_DEFAULT]: LOG_LEVEL.WARN,
      tag1: LOG_LEVEL.INFO,
    },
  })

  test('tag1 to pass', () => {
    const log = logService.getLogger()
    log.info('hello') // out
    log.warn('hello_warn') // pass
    log.tag('tag1').info('hello_tag1') // pass
    log.tag('tag1', 'tag2').info('hello_tag1') // out, cause all tags should pass
    expect(transportMock).toHaveBeenCalledTimes(2)
  })
})

describe('edge cases for coverage', () => {
  test('LoggerService with default opts', () => {
    new LoggerService()
  })

  test('ConsoleTransport', () => {
    const logService = new LoggerService({
      transports: [consoleTransport],
    })
    const log = logService.getLogger()
    log() // won't be called, 0 args
    log.log(LOG_LEVEL.INFO, 'hello') // called once
    log.tag().info() // won't be called, 0 args
    expect(consoleTransportMock).toHaveBeenCalledTimes(1)

    // for coverage
    consoleTransportMock.mockRestore()
    log('hello')
    log.log(LOG_LEVEL.DISABLED, 'hello')
  })

  test('noopFilter', () => {
    const logService = new LoggerService({
      transports: [noopTransport],
      filters: [noopFilter],
    })
    const log = logService.getLogger()
    log('hello')
    expect(noopFilterMock).toHaveBeenCalledTimes(1)

    jest.restoreAllMocks()
    log('hello') // for coverage
  })
})

function callBasicAPI (log: Logger): void {
  log('hello')
  log.debug('hello')
  log.info('hello')
  log.warn('hello')
  log.error('hello')
}
