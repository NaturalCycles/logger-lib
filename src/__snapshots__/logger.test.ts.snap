// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal logger with debug level basic api 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "INFO",
        "meta": undefined,
        "tags": Array [],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "DEBUG",
        "meta": undefined,
        "tags": Array [],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "INFO",
        "meta": undefined,
        "tags": Array [],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "WARN",
        "meta": undefined,
        "tags": Array [],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "ERROR",
        "meta": undefined,
        "tags": Array [],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`minimal logger with debug level basic api with tag1 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "INFO",
        "meta": undefined,
        "tags": Array [
          "tag1",
        ],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "DEBUG",
        "meta": undefined,
        "tags": Array [
          "tag1",
        ],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "INFO",
        "meta": undefined,
        "tags": Array [
          "tag1",
        ],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "WARN",
        "meta": undefined,
        "tags": Array [
          "tag1",
        ],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "ERROR",
        "meta": undefined,
        "tags": Array [
          "tag1",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`minimal logger with debug level deduplicate tags 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "INFO",
        "meta": undefined,
        "tags": Array [
          "tag1",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`minimal logger with default levels basic api 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "INFO",
        "meta": undefined,
        "tags": Array [],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "INFO",
        "meta": undefined,
        "tags": Array [],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "WARN",
        "meta": undefined,
        "tags": Array [],
      },
    ],
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "ERROR",
        "meta": undefined,
        "tags": Array [],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`minimal logger with default levels basic meta 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "args": Array [
          "hello",
        ],
        "level": "INFO",
        "meta": Object {
          "a": "a",
          "n": 1,
        },
        "tags": Array [],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
